src/VisualStudio/LiveShare/Impl/Shims/RunCodeActionsHandlerShim.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.LanguageServer.CustomProtocol;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.VisualStudio.LiveShare.LanguageServices;
+using Newtonsoft.Json.Linq;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.VisualStudio.LanguageServices.LiveShare
+{
+    /// <summary>
+    /// Run code actions handler.  Called when lightbulb invoked.
+    /// </summary>
+    [ExportLspRequestHandler(LiveShareConstants.RoslynContractName, LSP.Methods.WorkspaceExecuteCommandName)]
+    internal class RunCodeActionsHandlerShim : AbstractLiveShareHandlerShim<LSP.ExecuteCommandParams, object>
+    {
+        private readonly IThreadingContext _threadingContext;
+
+        [ImportingConstructor]
+        public RunCodeActionsHandlerShim([ImportMany] IEnumerable<Lazy<IRequestHandler, IRequestHandlerMetadata>> requestHandlers, IThreadingContext threadingContext)
+            : base(requestHandlers, LSP.Methods.WorkspaceExecuteCommandName)
+        {
+            _threadingContext = threadingContext;
+        }
+
+        /// <summary>
+        /// Unwraps the run code action request from the liveshare shape into the expected shape
+        /// handled by <see cref="RunCodeActionsHandler"/>.  This shape is an <see cref="LSP.Command"/>
+        /// containing <see cref="RunCodeActionParams"/> as the argument.
+        /// </summary>
+        public async override Task<object> HandleAsync(LSP.ExecuteCommandParams request, RequestContext<Solution> requestContext, CancellationToken cancellationToken)
+        {
+            // Unwrap the command to get the RunCodeActions command.
+            if (request.Command.StartsWith(CodeActionsHandlerShim.RemoteCommandNamePrefix))
```

Comment:
do we know why we need to do this? I guess protocol changed and we need this for old client?

Reply:
Not really - basically the liveshare code action implementation wrapped either a **command** or a **codeaction** inside a **command**.  This is because VSCode needed everything to be a command.

Basically from the codeaction method, you get 
```
command -> command
                 OR
                 -> codeaction -> workspace edit
```
So on the run code action, the client doesn't unwrap the command, just passes back to the server to execute.  So we have to unwrap it.

However, the spec for code actions mentions that it can return `(Command | CodeAction)[]`
https://microsoft.github.io/language-server-protocol/specification#textDocument_codeAction
So I made the base library handler just return the command/codeaction according to the spec, and just re-wrap it according to how liveshare expects in the shim